### PR TITLE
[Feature] Add SwAV

### PR DESCRIPTION
**Pull request by shiyutang**
### PR types

Features

### PR changes

Models

### Description

Reproduce SwAV on pre-train, fine-tune, linear probe.

Finetune-semi10%：
image

Linear probe:
image

pre-train:
image

### CODE REVIEW

File: passl/data/preprocess/basic_transforms.py [hunk 1]
```diff
@@ -941,3 +942,21 @@ def __call__(self, img):
             else:
                 img = ImageOps.solarize(img)
         return img
+
+
+class GaussianBlur(object):
```
Comment (GuoxiaWang) [inline passl/data/preprocess/basic_transforms.py:947]: `SimCLRGaussianBlur` 已经实现了

File: passl/engine/loops/classification_loop.py [hunk 2]
```diff
@@ -31,6 +31,91 @@
...
 
+
+import os
+import logging
+import time
+from datetime import timedelta
+import pandas as pd
+
+
+class LogFormatter:
```
Comment (GuoxiaWang) [inline passl/engine/loops/classification_loop.py:42]: Logger 是临时调试还是？logger 是已经有的

**Review by GuoxiaWang — CHANGES_REQUESTED** [review]

File: passl/engine/loops/classification_loop.py (hunk 2 above)
Comment (shiyutang) [reply passl/engine/loops/classification_loop.py:42]: 临时调试的

File: passl/data/preprocess/basic_transforms.py (hunk 1 above)
Comment (shiyutang) [reply passl/data/preprocess/basic_transforms.py]: don e

File: passl/engine/loops/classification_loop.py (hunk 2 above)
Comment (shiyutang) [reply passl/engine/loops/classification_loop.py]: removed

File: passl/models/swav_resnet.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+def kaiming_normal_init(param, **kwargs):
+    initializer = nn.initializer.KaimingNormal(**kwargs)
+    initializer(param, param.block)
+
+def constant_init(param, **kwargs):
+    initializer = nn.initializer.Constant(**kwargs)
+    initializer(param, param.block)
+
+
+class SwAVResNet(paddle.nn.Layer):
```
Comment (GuoxiaWang) [inline passl/models/swav_resnet.py:30]: 把这个文件的代码挪到 swav.py  中，删除 swav_resnet.py

File: passl/engine/loops/contrastive_learning_loop.py [hunk 4]
```diff
@@ -67,15 +62,20 @@ def forward_backward(self, batch):
             scaled = self.trainer.scaler.scale(loss_dict["loss"])
             scaled.backward()
 
+            try:
```
Comment (GuoxiaWang) [inline passl/engine/loops/contrastive_learning_loop.py:65]: 这个地方可以去掉，不用传入 total_iterations。 相关的变量从 `runtime_info_hub` 中获取，如果没有新增一下。

File: passl/engine/engine.py [hunk 5]
```diff
@@ -233,14 +233,14 @@ def worker_init_fn(worker_id):
...
         # load pretrained model
-        if self.config["Global"]["pretrained_model"] is not None:
-            assert isinstance(
-                self.config["Global"]["pretrained_model"], str
-            ), "pretrained_model type is not available. Please use `string`."
-            self.model.load_pretrained(
-                self.config["Global"]["pretrained_model"],
-                self.config["Global"]["rank"],
-                self.config["Global"].get("finetune", False))
+            if  self.config["Global"]["pretrained_model"] is not None:
```
Comment (GuoxiaWang) [inline passl/engine/engine.py:236]: 这里是什么原因导致有缩进

File: passl/data/dataset/imagefolder_dataset.py [hunk 6]
```diff
@@ -56,11 +59,25 @@ def __init__(self,
...
                  target_transform=None,
                  loader=default_loader,
-                 extensions=IMG_EXTENSIONS):
+                 extensions=IMG_EXTENSIONS,
+                 samples_tag=None):
 
         self.root = root
         classes, class_to_idx = self.find_classes(self.root)
-        samples = self.make_dataset(self.root, class_to_idx, extensions)
+        if samples_tag is None:
```
Comment (GuoxiaWang) [inline passl/data/dataset/imagefolder_dataset.py:67]: 这一些代码，是否可以在子类中重写？而不修改基类 `ImageFolder`

File: passl/engine/loops/loop.py [hunk 7]
```diff
@@ -219,7 +219,7 @@ def run(self):
                 self.trainer.train_dataloader.batch_sampler.set_epoch(epoch_id)
 
             # for one epoch train
-            self.train_one_epoch()
+            self.train_one_epoch(epoch_id)
```
Comment (GuoxiaWang) [inline passl/engine/loops/loop.py:222]: epoch_id 其实可以从 trainer 中获取，没必要传参数

File: passl/scheduler/__init__.py [hunk 8]
```diff
@@ -23,6 +24,11 @@ def build_lr_scheduler(lr_config, epochs, step_each_epoch):
     lr_config.update({'epochs': epochs, 'step_each_epoch': step_each_epoch})
     if 'name' in lr_config:
         lr_name = lr_config.pop('name')
+        if "MultiStepDecay" in lr_name:
```
Comment (GuoxiaWang) [inline passl/scheduler/__init__.py:27]: 其实这里没必要加这个 if，只要在 MultiStepDecay 中加 **kwargs 即可，不会有影响的。

**Review by GuoxiaWang — CHANGES_REQUESTED** [review]

File: passl/engine/loops/contrastive_learning_loop.py [hunk 9]
```diff
@@ -67,10 +62,14 @@ def forward_backward(self, batch):
...
+                self.trainer.model.after_loss_backward(self.total_iterations)
```
Comment (shiyutang) [inline passl/engine/loops/contrastive_learning_loop.py:66]: 在model的forward里面进行尝试

File: passl/engine/loops/loop.py [hunk 10]
```diff
@@ -264,7 +265,7 @@ def train_one_epoch(self):
 
         for batch_idx, batch in enumerate(self.trainer.train_dataloader):
             self.cur_batch_idx = batch_idx
-            self.total_iterations = (self.trainer.cur_epoch_id-1)*self.total_batch_idx + batch_idx
+            runtime_info_hub.total_iterations = (self.trainer.cur_epoch_id-1)*self.total_batch_idx + batch_idx
```
Comment (GuoxiaWang) [inline passl/engine/loops/loop.py:268]: 这个地方是不是等同于 global_steps?

File: passl/data/dataset/multicrop_dataset.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+                 num_crops,
+                 min_scale_crops,
+                 max_scale_crops):
+        super(MultiCropDataset, self).__init__(root)
+
+        assert len(size_crops) == len(num_crops)
+        assert len(min_scale_crops) == len(num_crops)
+        assert len(max_scale_crops) == len(num_crops)
+
+        color_transform = [get_color_distortion(), get_pil_gaussian_blur()]
```
Comment (GuoxiaWang) [inline passl/data/dataset/multicrop_dataset.py:45]: 为什么这个地方还有写死的操作？

File: passl/engine/loops/loop.py (hunk 10 above)
Comment (GuoxiaWang) [reply passl/engine/loops/loop.py:268]: 在 engine.py 里面

File: passl/data/dataset/fewshot_dataset.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+        assert txt_file_name is not None, "The txt_file_name should not be assigned."
+        if os.path.isfile(txt_file_name):
+            with open(txt_file_name, 'r') as f:
+                list_imgs = [li.split('\n')[0] for li in f.readlines()]
+
+            self.imgs = [(os.path.join(root, li.split('_')[0], li), self.class_to_idx[li.split('_')[0]]) for li in list_imgs]
+        else:
+            raise FileNotFoundError('{} is not existed'.format(txt_file_name))
+        print('Previous information is not correct.')
```
Comment (GuoxiaWang) [inline passl/data/dataset/fewshot_dataset.py:57]: print  改成 logger.info 吧

File: passl/models/swav.py [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+
+    def _freeze_norm(self, layer):
+        if isinstance(layer, (nn.layer.norm._BatchNormBase)):
+            layer._use_global_stats = True
+
+class SwAVLinearProbe(SwAV):
+    def __init__(self, class_num=1000, **kwargs):
+        super().__init__(**kwargs)
+        self.linear = RegLog(class_num)
+        self.res_model.eval()
```
Comment (GuoxiaWang) [inline passl/models/swav.py:93]: 这几个 res_model 起名是否可以换一个更通用的？比如 backbone 或者 encoder 也行，虽然SwAV 中就是resnet

File: passl/models/swav.py [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+    flags['FLAGS_cudnn_deterministic'] = False
+    paddle.set_flags(flags)
+
+    model = SwAVPretrain(**kwargs)
+
+    if paddle.distributed.get_world_size() > 1:
+        if not apex:
+            model = nn.SyncBatchNorm.convert_sync_batchnorm(model)
+        else:
+            # with apex syncbn speeds up computation than global syncbn
```
Comment (GuoxiaWang) [inline passl/models/swav.py:223]: 这个 apex 是否有必要？

File: passl/models/swav.py [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+            process_group = apex.parallel.create_syncbn_process_group(8)
+            model = apex.parallel.convert_syncbn_model(model, process_group=process_group)
+
+    return model
+
+class RegLog(paddle.nn.Layer):
```
Comment (GuoxiaWang) [inline passl/models/swav.py:229]: 这个 RegLog 改成 RegLogit 是不是会更清晰，我第一次读的时候以为是一个 log 相关的。

File: tasks/ssl/swav/README.md [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+      title={Unsupervised Learning of Visual Features by Contrasting Cluster Assignments},
+      author={Mathilde Caron and Ishan Misra and Julien Mairal and Priya Goyal and Piotr Bojanowski and Armand Joulin},
+      year={2021},
+      eprint={2006.09882},
+      archivePrefix={arXiv},
+      primaryClass={cs.CV}
+}
+```
+
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (GuoxiaWang) [inline tasks/ssl/swav/README.md:130]: README 中为什么会有这些代码？

File: tests/CI/case.sh [hunk 17]
```diff
@@ -399,9 +448,19 @@ function check_result() {
     fi
 
     echo -e "loss_base: $2 loss_test: $3" | tee -a $log_path/result.log
-    if [ $2 != $3 ];then
-      echo -e "\033 $1 loss diff check failed! \033" | tee -a $log_path/result.log
-      exit -1
+    diff=$(echo $2 $3|awk '{printf "%0.2f\n", ($2-$1)/$1*100}')
+    if [ $1 == swav_resnet50_224_pt_in1k_1n8c_dp_fp16o1 ];then
```
Comment (GuoxiaWang) [inline tests/CI/case.sh:452]: 这里专门为 swav_resnet50_224_pt_in1k_1n8c_dp_fp16o1 写了个分支，目的是什么？

File: tasks/ssl/swav/README.md (hunk 16 above)
Comment (shiyutang) [reply tasks/ssl/swav/README.md:130]: removed～

File: passl/models/swav.py (hunk 15 above)
Comment (shiyutang) [reply passl/models/swav.py:229]: 好的～

File: passl/models/swav.py (hunk 14 above)
Comment (shiyutang) [reply passl/models/swav.py:223]: 有可能会加速，但是还没有实验，可以保留用于后续实验。

File: passl/models/swav.py (hunk 13 above)
Comment (shiyutang) [reply passl/models/swav.py:93]: 可以改，但是训练的参数可能加载不上了

Comment (GuoxiaWang) [reply passl/models/swav.py:93]: 那先不改了把。有空可以改，不用再重新训练，只需要把 checkpoint 加载起来，换个key就行。

File: passl/data/dataset/fewshot_dataset.py (hunk 12 above)
Comment (shiyutang) [reply passl/data/dataset/fewshot_dataset.py:57]: done

File: tests/CI/case.sh (hunk 17 above)
Comment (shiyutang) [reply tests/CI/case.sh:452]: 因为单机八卡总是会差一点，没有办法固定所有权重。

File: passl/data/dataset/multicrop_dataset.py (hunk 11 above)
Comment (shiyutang) [reply passl/data/dataset/multicrop_dataset.py:45]: 这个数据集的功能就是实现SwAV提出的数据增强，感觉不是很需要配置的方式。

Comment (GuoxiaWang) [reply passl/data/dataset/multicrop_dataset.py:45]: 那在类名上加个前缀吧。这样就是特定的数据集。

File: passl/engine/loops/loop.py (hunk 10 above)
Comment (GuoxiaWang) [reply passl/engine/loops/loop.py:268]: 在 engine 中
```
    @property
    def global_step(self):
        return self.train_loop.global_step
```

File: passl/data/dataset/multicrop_dataset.py (hunk 11 above)
Comment (shiyutang) [reply passl/data/dataset/multicrop_dataset.py:45]: 好的～

**Review by GuoxiaWang — APPROVED** [review]
LGTM
